File: lua/test/codespace_entity_test.lua
-- Codespace entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("github_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("CodespaceEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Codespace(nil)
    assert.is_not_nil(ent)
  end)

  -- Feature #4: the entity stream(action, ...) method runs the op pipeline and
  -- returns an iterator over result items. With the streaming feature active it
  -- yields the feature's incremental output; otherwise it falls back to the
  -- materialised list so stream always yields.
  it("should stream", function()
    local seed = {
      entity = {
        ["codespace"] = {
          s1 = { id = "s1" },
          s2 = { id = "s2" },
          s3 = { id = "s3" },
        },
      },
    }

    -- Fallback: streaming inactive -> yields the materialised list items.
    local base = sdk.test(seed, nil)
    local seen = {}
    for item in base:Codespace(nil):stream("list", nil, nil) do
      table.insert(seen, item)
    end
    assert.are.equal(3, #seen)

    -- Inbound: streaming active -> yields each item from the feature.
    local config = require("config_shared")()
    if type(config.feature) == "table" and config.feature.streaming ~= nil then
      local streamsdk = sdk.test(seed, { feature = { streaming = { active = true } } })
      local got = {}
      for item in streamsdk:Codespace(nil):stream("list", nil, nil) do
        if vs.islist(item) then
          for _, sub in ipairs(item) do
            table.insert(got, sub)
          end
        else
          table.insert(got, item)
        end
      end
      assert.are.equal(3, #got)
    end
  end)

  it("should run basic flow", function()
    local setup = codespace_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "list", "update", "load", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "codespace." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_CODESPACE_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local codespace_ref01_ent = client:Codespace(nil)
    local codespace_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.codespace"), "codespace_ref01"))
    codespace_ref01_data["codespace_id"] = setup.idmap["codespace01"]
    codespace_ref01_data["codespace_name"] = setup.idmap["codespace_name01"]
    codespace_ref01_data["member_id"] = setup.idmap["member01"]
    codespace_ref01_data["org_id"] = setup.idmap["org01"]
    codespace_ref01_data["owner"] = setup.idmap["owner01"]
    codespace_ref01_data["repo"] = setup.idmap["repo01"]
    codespace_ref01_data["secret_id"] = setup.idmap["secret01"]
    codespace_ref01_data["secret_name"] = setup.idmap["secret_name01"]
    codespace_ref01_data["username"] = setup.idmap["username01"]

    local codespace_ref01_data_result, err = codespace_ref01_ent:create(codespace_ref01_data, nil)
    assert.is_nil(err)
    codespace_ref01_data = helpers.to_map(type(codespace_ref01_data_result) == 'table' and codespace_ref01_data_result.data_get and codespace_ref01_data_result:data_get() or codespace_ref01_data_result)
    assert.is_not_nil(codespace_ref01_data)
    assert.is_not_nil(codespace_ref01_data["id"])

    -- LIST
    local codespace_ref01_match = {
      ["secret_name"] = setup.idmap["secret_name01"],
    }

    local codespace_ref01_list_result, err = codespace_ref01_ent:list(codespace_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(codespace_ref01_list_result)

    local found_item = vs.select(
      runner.entity_list_to_data(codespace_ref01_list_result),
      { id = codespace_ref01_data["id"] })
    assert.is_false(vs.isempty(found_item))

    -- UPDATE
    local codespace_ref01_data_up0_up = {
      id = codespace_ref01_data["id"],
    }

    local codespace_ref01_markdef_up0_name = "archive_url"
    local codespace_ref01_markdef_up0_value = "Mark01-codespace_ref01_" .. tostring(setup.now)
    codespace_ref01_data_up0_up[codespace_ref01_markdef_up0_name] = codespace_ref01_markdef_up0_value

    local codespace_ref01_resdata_up0_result, err = codespace_ref01_ent:update(codespace_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local codespace_ref01_resdata_up0 = helpers.to_map(type(codespace_ref01_resdata_up0_result) == 'table' and codespace_ref01_resdata_up0_result.data_get and codespace_ref01_resdata_up0_result:data_get() or codespace_ref01_resdata_up0_result)
    assert.is_not_nil(codespace_ref01_resdata_up0)
    assert.are.equal(codespace_ref01_resdata_up0["id"], codespace_ref01_data_up0_up["id"])
    assert.are.equal(codespace_ref01_resdata_up0[codespace_ref01_markdef_up0_name], codespace_ref01_markdef_up0_value)

    -- LOAD
    local codespace_ref01_match_dt0 = {
      id = codespace_ref01_data["id"],
    }
    local codespace_ref01_data_dt0_loaded, err = codespace_ref01_ent:load(codespace_ref01_match_dt0, nil)
    assert.is_nil(err)
    local codespace_ref01_data_dt0_load_result = helpers.to_map(type(codespace_ref01_data_dt0_loaded) == 'table' and codespace_ref01_data_dt0_loaded.data_get and codespace_ref01_data_dt0_loaded:data_get() or codespace_ref01_data_dt0_loaded)
    assert.is_not_nil(codespace_ref01_data_dt0_load_result)
    assert.are.equal(codespace_ref01_data_dt0_load_result["id"], codespace_ref01_data["id"])

    -- REMOVE
    local codespace_ref01_match_rm0 = {
      id = codespace_ref01_data["id"],
    }
    local _, err = codespace_ref01_ent:remove(codespace_ref01_match_rm0, nil)
    assert.is_nil(err)

    -- LIST
    local codespace_ref01_match_rt0 = {
      ["secret_name"] = setup.idmap["secret_name01"],
    }

    local codespace_ref01_list_rt0_result, err = codespace_ref01_ent:list(codespace_ref01_match_rt0, nil)
    assert.is_nil(err)
    assert.is_table(codespace_ref01_list_rt0_result)

    local not_found_item = vs.select(
      runner.entity_list_to_data(codespace_ref01_list_rt0_result),
      { id = codespace_ref01_data["id"] })
    assert.is_true(vs.isempty(not_found_item))

  end)
end)

function codespace_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/codespace/CodespaceTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read codespace test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "codespace01", "codespace02", "codespace03", "org01", "org02", "org03", "repo01", "repo02", "repo03", "secret01", "secret02", "secret03", "member01", "member02", "member03", "pull01", "pull02", "pull03", "export01", "export02", "export03", "repository01", "repository02", "repository03", "codespace_name01", "owner01", "secret_name01", "username01" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("GITHUB_TEST_CODESPACE_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["GITHUB_TEST_CODESPACE_ENTID"] = idmap,
    ["GITHUB_TEST_LIVE"] = "FALSE",
    ["GITHUB_TEST_EXPLAIN"] = "FALSE",
    ["GITHUB_APIKEY"] = "",
  })

  local idmap_resolved = helpers.to_map(
    env["GITHUB_TEST_CODESPACE_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["GITHUB_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      -- FIRST, so the generated fields below win: sdk-test-control.json's
      -- test.client.options adds to the live client, it does not redirect it.
      runner.live_client_options(),
      {
        apikey = env["GITHUB_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["GITHUB_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["GITHUB_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: go/test/codespace_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/github-sdk/go"
	"github.com/voxgig-sdk/github-sdk/go/core"

	vs "github.com/voxgig-sdk/github-sdk/go/utility/struct"
)

func TestCodespaceEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Codespace(nil)
		if ent == nil {
			t.Fatal("expected non-nil CodespaceEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"codespace": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Codespace(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Codespace(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := codespaceBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "codespace." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_CODESPACE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		codespaceRef01Ent := client.Codespace(nil)
		codespaceRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "codespace"}), "codespace_ref01"))
		codespaceRef01Data["codespace_id"] = setup.idmap["codespace01"]
		codespaceRef01Data["codespace_name"] = setup.idmap["codespace_name01"]
		codespaceRef01Data["member_id"] = setup.idmap["member01"]
		codespaceRef01Data["org_id"] = setup.idmap["org01"]
		codespaceRef01Data["owner"] = setup.idmap["owner01"]
		codespaceRef01Data["repo"] = setup.idmap["repo01"]
		codespaceRef01Data["secret_id"] = setup.idmap["secret01"]
		codespaceRef01Data["secret_name"] = setup.idmap["secret_name01"]
		codespaceRef01Data["username"] = setup.idmap["username01"]

		codespaceRef01DataResult, err := codespaceRef01Ent.Create(codespaceRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		codespaceRef01Data = core.ToMapAny(entityData(codespaceRef01DataResult))
		if codespaceRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if codespaceRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		codespaceRef01Match := map[string]any{
			"secret_name": setup.idmap["secret_name01"],
		}

		codespaceRef01ListResult, err := codespaceRef01Ent.List(codespaceRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		codespaceRef01List, codespaceRef01ListOk := codespaceRef01ListResult.([]any)
		if !codespaceRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", codespaceRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(codespaceRef01List), map[string]any{"id": codespaceRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		codespaceRef01DataUp0Up := map[string]any{
			"id": codespaceRef01Data["id"],
		}

		codespaceRef01MarkdefUp0Name := "archive_url"
		codespaceRef01MarkdefUp0Value := fmt.Sprintf("Mark01-codespace_ref01_%d", setup.now)
		codespaceRef01DataUp0Up[codespaceRef01MarkdefUp0Name] = codespaceRef01MarkdefUp0Value

		codespaceRef01ResdataUp0Result, err := codespaceRef01Ent.Update(codespaceRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		codespaceRef01ResdataUp0 := core.ToMapAny(entityData(codespaceRef01ResdataUp0Result))
		if codespaceRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if codespaceRef01ResdataUp0["id"] != codespaceRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if codespaceRef01ResdataUp0[codespaceRef01MarkdefUp0Name] != codespaceRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", codespaceRef01MarkdefUp0Name, codespaceRef01ResdataUp0[codespaceRef01MarkdefUp0Name])
		}

		// LOAD
		codespaceRef01MatchDt0 := map[string]any{
			"id": codespaceRef01Data["id"],
		}
		codespaceRef01DataDt0Loaded, err := codespaceRef01Ent.Load(codespaceRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		codespaceRef01DataDt0LoadResult := core.ToMapAny(entityData(codespaceRef01DataDt0Loaded))
		if codespaceRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if codespaceRef01DataDt0LoadResult["id"] != codespaceRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		codespaceRef01MatchRm0 := map[string]any{
			"id": codespaceRef01Data["id"],
		}
		_, err = codespaceRef01Ent.Remove(codespaceRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		codespaceRef01MatchRt0 := map[string]any{
			"secret_name": setup.idmap["secret_name01"],
		}

		codespaceRef01ListRt0Result, err := codespaceRef01Ent.List(codespaceRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		codespaceRef01ListRt0, codespaceRef01ListRt0Ok := codespaceRef01ListRt0Result.([]any)
		if !codespaceRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", codespaceRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(codespaceRef01ListRt0), map[string]any{"id": codespaceRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func codespaceBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "codespace", "CodespaceTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read codespace test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse codespace test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"codespace01", "codespace02", "codespace03", "org01", "org02", "org03", "repo01", "repo02", "repo03", "secret01", "secret02", "secret03", "member01", "member02", "member03", "pull01", "pull02", "pull03", "export01", "export02", "export03", "repository01", "repository02", "repository03", "codespace_name01", "owner01", "secret_name01", "username01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITHUB_TEST_CODESPACE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITHUB_TEST_CODESPACE_ENTID": idmap,
		"GITHUB_TEST_LIVE":      "FALSE",
		"GITHUB_TEST_EXPLAIN":   "FALSE",
		"GITHUB_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["GITHUB_TEST_CODESPACE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["GITHUB_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["GITHUB_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewGithubSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITHUB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITHUB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
